Bundling wrong file initially when doing component tests

Cypress v4.5.0

I used https://github.com/bahmutov/cypress-react-unit-test/ and ran

```
DEBUG=cypress:webpack npm run cy:open
```

and noticed when clicking on a component spec that it is asking to bundle `GET /__cypress/iframes/component/basic/counter-set-state/counter-spec.js` but then I see the following error messages

```
  cypress:webpack get /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js +16s
  cypress:webpack webpackOptions: { resolve: { alias: { react: '/Users/gleb/git/cypress-react-unit-test/node_modules/react' } }, mode: 'development', stats: 'verbose', module: { rules: [ [Object], [Object], [Object] ] } } +1ms
  cypress:webpack watchOptions: {} +0ms
  cypress:webpack input: /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js +1ms
  cypress:webpack output: /Users/gleb/Library/Application Support/Cypress/cy/production/projects/cypress-react-unit-test-3d4057e937d798d828e0b355d0057df4/bundles/cypress/integration/component/basic/counter-set-state/counter-spec.js +0ms
  cypress:webpack watching +145ms
  cypress:webpack compile /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js +2ms
  cypress:webpack get /Users/gleb/git/cypress-react-unit-test/cypress/support/index.js +17ms
  cypress:webpack webpackOptions: { resolve: { alias: { react: '/Users/gleb/git/cypress-react-unit-test/node_modules/react' } }, mode: 'development', stats: 'verbose', module: { rules: [ [Object], [Object], [Object] ] }, entry: [ '/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js' ], output: { path: '/Users/gleb/Library/Application Support/Cypress/cy/production/projects/cypress-react-unit-test-3d4057e937d798d828e0b355d0057df4/bundles/cypress/integration/component/basic/counter-set-state', filename: 'counter-spec.js' }, devtool: 'inline-source-map' } +0ms
  cypress:webpack watchOptions: {} +1ms
  cypress:webpack input: /Users/gleb/git/cypress-react-unit-test/cypress/support/index.js +0ms
  cypress:webpack output: /Users/gleb/Library/Application Support/Cypress/cy/production/projects/cypress-react-unit-test-3d4057e937d798d828e0b355d0057df4/bundles/cypress/support/index.js +0ms
  cypress:webpack watching +4ms
  cypress:webpack compile /Users/gleb/git/cypress-react-unit-test/cypress/support/index.js +0ms
  cypress:webpack get /Users/gleb/git/cypress-react-unit-test/cypress/component/basic/counter-set-state/counter-spec.js +5ms
  cypress:webpack webpackOptions: { resolve: { alias: { react: '/Users/gleb/git/cypress-react-unit-test/node_modules/react' } }, mode: 'development', stats: 'verbose', module: { rules: [ [Object], [Object], [Object] ] }, entry: [ '/Users/gleb/git/cypress-react-unit-test/cypress/support/index.js' ], output: { path: '/Users/gleb/Library/Application Support/Cypress/cy/production/projects/cypress-react-unit-test-3d4057e937d798d828e0b355d0057df4/bundles/cypress/support', filename: 'index.js' }, devtool: 'inline-source-map' } +0ms
  cypress:webpack watchOptions: {} +0ms
  cypress:webpack input: /Users/gleb/git/cypress-react-unit-test/cypress/component/basic/counter-set-state/counter-spec.js +0ms
  cypress:webpack output: /Users/gleb/Library/Application Support/Cypress/cy/production/projects/cypress-react-unit-test-3d4057e937d798d828e0b355d0057df4/bundles/cypress/component/basic/counter-set-state/counter-spec.js +1ms
  cypress:webpack watching +2ms
  cypress:webpack compile /Users/gleb/git/cypress-react-unit-test/cypress/component/basic/counter-set-state/counter-spec.js +0ms
  cypress:webpack stats had error(s) +138ms
  cypress:webpack errored bundling /Users/gleb/Library/Application Support/Cypress/cy/production/projects/cypress-react-unit-test-3d4057e937d798d828e0b355d0057df4/bundles/cypress/integration/component/basic/counter-set-state/counter-spec.js Webpack Compilation Error
multi ./cypress/integration/component/basic/counter-set-state/counter-spec.js
Module not found: Error: Can't resolve '/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js' in '/Users/gleb/git/cypress-react-unit-test'
resolve '/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js' in '/Users/gleb/git/cypress-react-unit-test'
  using description file: /Users/gleb/git/cypress-react-unit-test/package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: /Users/gleb/git/cypress-react-unit-test/package.json (relative path: ./cypress/integration/component/basic/counter-set-state/counter-spec.js)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js doesn't exist
      .wasm
        Field 'browser' doesn't contain a valid alias configuration
        /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.wasm doesn't exist
      .mjs
        Field 'browser' doesn't contain a valid alias configuration
        /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.mjs doesn't exist
      .js
        Field 'browser' doesn't contain a valid alias configuration
        /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.js doesn't exist
      .json
        Field 'browser' doesn't contain a valid alias configuration
        /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.json doesn't exist
      as directory
        /Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js doesn't exist
[/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js]
[/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.wasm]
[/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.mjs]
[/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.js]
[/Users/gleb/git/cypress-react-unit-test/cypress/integration/component/basic/counter-set-state/counter-spec.js.json]
 @ multi ./cypress/integration/component/basic/counter-set-state/counter-spec.js main[0] +1ms
```

and then it is asking to bundle again

```
GET /__cypress/tests?p=cypress/component/basic/counter-set-state/counter-spec.js-250 - - ms - -
GET /__cypress/iframes/component/basic/counter-set-state/counter-spec.js 200 0.918 ms - 787
GET /__cypress/runner/popper.js.map 404 0.649 ms - 136
  cypress:webpack get /Users/gleb/git/cypress-react-unit-test/cypress/support/index.js +409ms
  cypress:webpack already have bundle for /Users/gleb/git/cypress-react-unit-test/cypress/support/index.js +0ms
  cypress:webpack get /Users/gleb/git/cypress-react-unit-test/cypress/component/basic/counter-set-state/counter-spec.js +1ms
  cypress:webpack already have bundle for /Users/gleb/git/cypress-react-unit-test/cypress/component/basic/counter-set-state/counter-spec.js +0ms
GET /__cypress/tests?p=cypress/support/index.js-864 200 302.270 ms - -
```

So something is wrong there. It is asking for file `./cypress/integration/component/basic/counter-set-state/counter-spec.js` and it should be `./cypress/component/basic/counter-set-state/counter-spec.js` (no "integration" folder)

--- a/packages/runner/src/iframe/iframes.jsx
+++ b/packages/runner/src/iframe/iframes.jsx
@@ -102,7 +102,7 @@ export default class Iframes extends Component {
     logger.clearLog()
     this._setScriptError(null)
 
-    this.props.eventManager.setup(config, specPath)
+    this.props.eventManager.setup(config)
 
     const $autIframe = this._loadIframes(specPath)
 

--- a/packages/runner/src/lib/event-manager.js
+++ b/packages/runner/src/lib/event-manager.js
@@ -189,7 +189,7 @@ const eventManager = {
     }
   },
 
-  setup (config, specPath) {
+  setup (config) {
     Cypress = this.Cypress = $Cypress.create(config)
 
     // expose Cypress globally
@@ -197,7 +197,7 @@ const eventManager = {
 
     this._addListeners()
 
-    ws.emit('watch:test:file', specPath)
+    ws.emit('watch:test:file', config.spec)
   },
 
   isBrowser (browserName) {

--- a/packages/server/lib/controllers/files.coffee
+++ b/packages/server/lib/controllers/files.coffee
@@ -31,11 +31,13 @@ module.exports = {
     .then (specs) =>
       @getJavascripts(config)
       .then (js) =>
-        res.render iframePath, {
+        iframeOptions = {
           title: @getTitle(test)
           domain: getRemoteState().domainName
           scripts:  JSON.stringify(js.concat(specs))
         }
+        debug("iframe %s options %o", test, iframeOptions)
+        res.render iframePath, iframeOptions
 
   getSpecs: (spec, config) ->
     debug("get specs %o", { spec })

--- a/packages/server/lib/socket.js
+++ b/packages/server/lib/socket.js
@@ -69,24 +69,19 @@ class Socket {
     })
   }
 
-  // TODO: clean this up by sending the spec object instead of
-  // the url path
-  watchTestFileByPath (config, originalFilePathOrSpecConfig, options) {
-    let filePath
+  watchTestFileByPath (config, specConfig, options) {
+    debug('watching spec with config %o', specConfig)
 
-    if (typeof originalFilePathOrSpecConfig === 'string') {
-      // files are always sent as integration/foo_spec.js
-      // need to take into account integrationFolder may be different so
-      // integration/foo_spec.js becomes cypress/my-integration-folder/foo_spec.js
-      debug('watch test file %o', originalFilePathOrSpecConfig)
-      const originalFilePath = originalFilePathOrSpecConfig
+    const cleanIntegrationPrefix = (s) => {
+      const removedIntegrationPrefix = path.join(config.integrationFolder, s.replace(`integration${path.sep}`, ''))
 
-      filePath = path.join(config.integrationFolder, originalFilePath.replace(`integration${path.sep}`, ''))
-      filePath = path.relative(config.projectRoot, filePath)
-    } else {
-      debug('watching spec with config %o', originalFilePathOrSpecConfig)
-      filePath = originalFilePathOrSpecConfig.absolute
+      return path.relative(config.projectRoot, removedIntegrationPrefix)
     }
+
+    // previously we have assumed that we pass integration spec path with "integration/" prefix
+    // now we pass spec config object that tells what kind of spec it is, has relative path already
+    // so the only special handling remains for special paths like "integration/__all"
+    const filePath = typeof specConfig === 'string' ? cleanIntegrationPrefix(specConfig) : specConfig.relative
 
     // bail if this is special path like "__all"
     // maybe the client should not ask to watch non-spec files?

--- a/packages/server/lib/socket.js
+++ b/packages/server/lib/socket.js
@@ -71,14 +71,22 @@ class Socket {
 
   // TODO: clean this up by sending the spec object instead of
   // the url path
-  watchTestFileByPath (config, originalFilePath, options) {
-    // files are always sent as integration/foo_spec.js
-    // need to take into account integrationFolder may be different so
-    // integration/foo_spec.js becomes cypress/my-integration-folder/foo_spec.js
-    debug('watch test file %o', originalFilePath)
-    let filePath = path.join(config.integrationFolder, originalFilePath.replace(`integration${path.sep}`, ''))
+  watchTestFileByPath (config, originalFilePathOrSpecConfig, options) {
+    let filePath
 
-    filePath = path.relative(config.projectRoot, filePath)
+    if (typeof originalFilePathOrSpecConfig === 'string') {
+      // files are always sent as integration/foo_spec.js
+      // need to take into account integrationFolder may be different so
+      // integration/foo_spec.js becomes cypress/my-integration-folder/foo_spec.js
+      debug('watch test file %o', originalFilePathOrSpecConfig)
+      const originalFilePath = originalFilePathOrSpecConfig
+
+      filePath = path.join(config.integrationFolder, originalFilePath.replace(`integration${path.sep}`, ''))
+      filePath = path.relative(config.projectRoot, filePath)
+    } else {
+      debug('watching spec with config %o', originalFilePathOrSpecConfig)
+      filePath = originalFilePathOrSpecConfig.absolute
+    }
 
     // bail if this is special path like "__all"
     // maybe the client should not ask to watch non-spec files?
@@ -287,8 +295,10 @@ class Socket {
         return options.onSpecChanged(spec)
       })
 
-      socket.on('watch:test:file', (filePath, cb = function () {}) => {
-        this.watchTestFileByPath(config, filePath, options)
+      socket.on('watch:test:file', (specInfo, cb = function () { }) => {
+        debug('watch:test:file %o', specInfo)
+
+        this.watchTestFileByPath(config, specInfo, options)
 
         // callback is only for testing purposes
         return cb()

--- a/packages/server/test/unit/socket_spec.coffee
+++ b/packages/server/test/unit/socket_spec.coffee
@@ -323,11 +323,12 @@ describe "lib/socket", ->
           done()
 
     context "on(watch:test:file)", ->
-      it "calls socket#watchTestFileByPath with config, filePath", (done) ->
+      it "calls socket#watchTestFileByPath with config, spec argument", (done) ->
         sinon.stub(@socket, "watchTestFileByPath")
 
-        @client.emit "watch:test:file", "path/to/file", =>
-          expect(@socket.watchTestFileByPath).to.be.calledWith(@cfg, "path/to/file")
+        specArgument = {}
+        @client.emit "watch:test:file", specArgument, =>
+          expect(@socket.watchTestFileByPath).to.be.calledWith(@cfg, specArgument)
           done()
 
     context "on(app:connect)", ->
@@ -497,6 +498,15 @@ describe "lib/socket", ->
       it "watches file by path", ->
         @socket.watchTestFileByPath(@cfg, "integration#{path.sep}test2.coffee")
         expect(preprocessor.getFile).to.be.calledWith("tests#{path.sep}test2.coffee", @cfg)
+
+      it "watches file by relative path in spec object", ->
+        # this is what happens now with component / integration specs
+        spec = {
+          absolute: "#{path.sep}foo#{path.sep}bar",
+          relative: "relative#{path.sep}to#{path.sep}root#{path.sep}test2.coffee"
+        }
+        @socket.watchTestFileByPath(@cfg, spec)
+        expect(preprocessor.getFile).to.be.calledWith(spec.relative, @cfg)
 
       it "triggers watched:file:changed event when preprocessor 'file:updated' is received", (done) ->
         sinon.stub(fs, "statAsync").resolves()

